File: zig/src/core/filter.zig
const std = @import("std");
const io = @import("io.zig");

pub const Filter = struct {
    allocator: std.mem.Allocator,
    ignore_patterns: std.ArrayList([]const u8),
    skip_hidden: bool = true,

    const builtin_skip_dirs = [_][]const u8{
        ".git", ".zig-cache", "zig-out", "node_modules", "target",
        ".mypy_cache", "__pycache__", ".tox", ".pytest_cache",
        "vendor", "dist", "build", ".next", ".nuxt",
        ".svelte-kit", ".parcel-cache", ".turbo",
        "coverage", ".nyc_output", ".cache",
        // Package / dependency caches that are not hidden (hidden ones like
        // .cargo/.rustup/.npm/.m2/.gradle/.venv are already skipped by skip_hidden).
        "venv", "site-packages", "Pods", "_build", "elm-stuff", "bower_components",
        // Terraform/IaC + build-image artifacts (mkosi builds a full OS rootfs
        // under mkosi.tools — thousands of system files that aren't project code).
        ".terraform", "mkosi.tools", "mkosi.cache", "mkosi.output", ".mkosi",
    };

    // Path fragments matched anywhere in a path. Used to prune language package
    // caches that live under non-hidden roots (e.g. ~/go/pkg/mod) which a
    // per-component match cannot catch without over-matching a dir literally
    // named "go" or "pkg".
    const builtin_skip_fragments = [_][]const u8{
        "go/pkg/mod", "go/pkg/sumdb", "go/pkg/cache",
    };

    const builtin_skip_extensions = [_][]const u8{
        ".exe", ".dll", ".so", ".dylib", ".o", ".a", ".lib",
        ".pyc", ".pyo", ".class", ".jar", ".war",
        ".wasm", ".min.js", ".min.css",
        ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".ico", ".svg",
        ".mp3", ".mp4", ".avi", ".mov", ".webm",
        ".zip", ".tar", ".gz", ".bz2", ".xz", ".rar", ".7z",
        ".pdf", ".doc", ".docx", ".xls", ".xlsx",
        ".sqlite", ".db", ".lock",
    };

    pub fn init(allocator: std.mem.Allocator) Filter {
        return .{
            .allocator = allocator,
            .ignore_patterns = std.ArrayList([]const u8).empty,
        };
    }

    pub fn deinit(self: *Filter) void {
        for (self.ignore_patterns.items) |p| self.allocator.free(p);
        self.ignore_patterns.deinit(self.allocator);
    }

    pub fn load_gitignore(self: *Filter, root: []const u8) !void {
        const gitignore_path = try std.fs.path.join(self.allocator, &.{ root, ".gitignore" });
        defer self.allocator.free(gitignore_path);

        const content = io.readFileAlloc(self.allocator, gitignore_path, 1024 * 1024) catch return;
        defer self.allocator.free(content);

        var line_it = std.mem.splitScalar(u8, content, '\n');
        while (line_it.next()) |line| {
            const trimmed = std.mem.trim(u8, line, " \t\r");
            if (trimmed.len == 0 or trimmed[0] == '#') continue;
            var pat = trimmed;
            if (pat.len > 0 and pat[pat.len - 1] == '/') pat = pat[0 .. pat.len - 1];
            if (pat.len == 0) continue;
            try self.ignore_patterns.append(self.allocator, try self.allocator.dupe(u8, pat));
        }
    }

    /// Whether traversal should skip descending into this directory.
    /// Reuses the same rules as should_ignore minus the extension check.
    pub fn should_skip_dir(self: *const Filter, path: []const u8) bool {
        for (&builtin_skip_fragments) |frag| {
            if (std.mem.indexOf(u8, path, frag) != null) return true;
        }
        var comp_it = std.mem.splitScalar(u8, path, '/');
        while (comp_it.next()) |component| {
            if (component.len == 0) continue;
            if (self.skip_hidden and component[0] == '.' and component.len > 1) return true;
            for (&builtin_skip_dirs) |dir| {
                if (std.mem.eql(u8, component, dir)) return true;
            }
        }
        for (self.ignore_patterns.items) |pattern| {
            if (matchPattern(path, pattern)) return true;
        }
        return false;
    }

    pub fn should_ignore(self: *const Filter, path: []const u8) bool {
        for (&builtin_skip_fragments) |frag| {
            if (std.mem.indexOf(u8, path, frag) != null) return true;
        }
        var comp_it = std.mem.splitScalar(u8, path, '/');
        while (comp_it.next()) |component| {
            if (component.len == 0) continue;

            // Skip hidden files/dirs
            if (self.skip_hidden and component[0] == '.' and component.len > 1) {
                if (!std.mem.eql(u8, component, ".gitignore") and
                    !std.mem.eql(u8, component, ".env") and
                    !std.mem.eql(u8, component, ".env.example"))
                {
                    return true;
                }
            }

            // Builtin skip dirs
            for (&builtin_skip_dirs) |dir| {
                if (std.mem.eql(u8, component, dir)) return true;
            }
        }

        // Builtin skip extensions
        for (&builtin_skip_extensions) |ext| {
            if (std.mem.endsWith(u8, path, ext)) return true;
        }

        // Gitignore patterns
        for (self.ignore_patterns.items) |pattern| {
            if (matchPattern(path, pattern)) return true;
        }

        return false;
    }
};

fn matchPattern(path: []const u8, pattern: []const u8) bool {
    if (pattern.len == 0) return false;
    if (pattern[0] == '!') return false;

    if (pattern[0] == '/') {
        return matchGlob(path, pattern[1..]);
    }

    if (matchGlob(path, pattern)) return true;

    const basename = std.fs.path.basename(path);
    if (matchGlob(basename, pattern)) return true;

    var it = std.mem.splitScalar(u8, path, '/');
    while (it.next()) |component| {
        if (std.mem.eql(u8, component, pattern)) return true;
    }

    return false;
}

fn matchGlob(str: []const u8, pattern: []const u8) bool {
    var si: usize = 0;
    var pi: usize = 0;
    var star_pi: ?usize = null;
    var star_si: usize = 0;

    while (si < str.len) {
        if (pi < pattern.len and (pattern[pi] == str[si] or pattern[pi] == '?')) {
            si += 1;
            pi += 1;
        } else if (pi < pattern.len and pattern[pi] == '*') {
            star_pi = pi;
            star_si = si;
            pi += 1;
        } else if (star_pi) |sp| {
            pi = sp + 1;
            star_si += 1;
            si = star_si;
        } else {
            return false;
        }
    }

    while (pi < pattern.len and pattern[pi] == '*') pi += 1;
    return pi == pattern.len;
}

File: decommissioned/src/filter.rs
use ignore::WalkBuilder;
use std::collections::HashSet;
use std::path::{Path, PathBuf};
use std::sync::Arc;

use crate::config::CodeIndexerConfig;

/// Built-in directories to always skip.
const SKIP_DIRS: &[&str] = &[
    ".git",
    ".hg",
    ".svn",
    ".bzr",
    "node_modules",
    ".next",
    ".nuxt",
    ".svelte-kit",
    "target",
    ".cargo",
    "__pycache__",
    ".pytest_cache",
    ".mypy_cache",
    ".ruff_cache",
    ".venv",
    "venv",
    "env",
    ".env",
    "vendor",
    ".bundle",
    "dist",
    "build",
    "out",
    ".idea",
    ".vscode",
    "coverage",
    ".nyc_output",
    ".tox",
    ".nox",
    ".terraform",
    ".cache",
    "bower_components",
    ".eggs",
    "*.egg-info",
    ".lock",
];

/// Built-in file extensions to skip (binary / non-text).
const SKIP_EXTENSIONS: &[&str] = &[
    "png", "jpg", "jpeg", "gif", "bmp", "ico", "webp", "svg", "tiff", "psd", "mp3", "mp4", "avi",
    "mov", "wmv", "flv", "mkv", "webm", "zip", "tar", "gz", "bz2", "xz", "7z", "rar", "tgz", "exe",
    "dll", "so", "dylib", "a", "lib", "o", "obj", "wasm", "class", "pyc", "pyo", "pdf", "doc",
    "docx", "xls", "xlsx", "ppt", "pptx", "woff", "woff2", "ttf", "eot", "otf", "lock", "sum",
];

/// File filter and directory walker that respects .gitignore and skips
/// binary/large/symlink-escape files.
pub struct FileFilter {
    config: Arc<CodeIndexerConfig>,
    skip_dirs: HashSet<String>,
    skip_extensions: HashSet<String>,
}

impl FileFilter {
    pub fn new(config: Arc<CodeIndexerConfig>) -> Self {
        let mut skip_dirs: HashSet<String> = SKIP_DIRS.iter().map(|s| s.to_string()).collect();
        skip_dirs.extend(config.extra_skip_dirs.iter().cloned());

        let mut skip_extensions: HashSet<String> =
            SKIP_EXTENSIONS.iter().map(|s| s.to_string()).collect();
        skip_extensions.extend(config.extra_skip_extensions.iter().cloned());

        Self {
            config,
            skip_dirs,
            skip_extensions,
        }
    }

    /// Build a `WalkBuilder` configured for the workspace root.
    /// Respects .gitignore, skips hidden files, never follows symlinks.
    pub fn walk_builder(&self) -> WalkBuilder {
        let mut builder = WalkBuilder::new(&self.config.workspace_root);
        builder
            .git_ignore(self.config.respect_gitignore)
            .git_exclude(self.config.respect_git_exclude)
            .git_global(self.config.respect_git_global)
            .hidden(self.config.skip_hidden)
            .follow_links(self.config.follow_symlinks)
            .max_filesize(Some(self.config.max_file_size));

        // Custom skip dirs via filter_entry
        let skip_dirs = self.skip_dirs.clone();
        builder.filter_entry(move |entry| {
            if entry.file_type().map_or(false, |ft| ft.is_dir()) {
                let name = entry.file_name().to_string_lossy();
                !skip_dirs.contains(name.as_ref())
            } else {
                true
            }
        });

        builder
    }

    /// Check if a file extension should be skipped.
    pub fn should_skip_extension(&self, path: &Path) -> bool {
        path.extension()
            .and_then(|e| e.to_str())
            .map(|ext| self.skip_extensions.contains(ext))
            .unwrap_or(false)
    }

    /// Detect if a file is binary by scanning for null bytes in the first 8KB.
    pub fn is_binary(path: &Path) -> bool {
        let Ok(data) = std::fs::read(path) else {
            return true;
        };
        let scan_len = data.len().min(8192);
        data[..scan_len].contains(&0)
    }

    /// Check if a resolved path escapes the workspace root.
    pub fn is_symlink_escape(path: &Path, workspace: &Path) -> bool {
        if let Ok(resolved) = path.canonicalize() {
            !resolved.starts_with(workspace)
        } else {
            false
        }
    }

    /// Collect all indexable file paths from the workspace.
    /// Returns paths relative to the workspace root.
    pub fn collect_paths(&self) -> Vec<PathBuf> {
        let workspace = &self.config.workspace_root;
        let mut paths = Vec::new();

        let walker = self.walk_builder();
        for result in walker.build() {
            let Ok(entry) = result else { continue };
            if !entry.file_type().map_or(false, |ft| ft.is_file()) {
                continue;
            }

            let path = entry.path();

            // Extension skip
            if self.should_skip_extension(path) {
                continue;
            }

            // Binary detection
            if Self::is_binary(path) {
                continue;
            }

            // Symlink escape check
            if !self.config.follow_symlinks && Self::is_symlink_escape(path, workspace) {
                continue;
            }

            // Make relative to workspace
            if let Ok(rel) = path.strip_prefix(workspace) {
                paths.push(rel.to_path_buf());
            }
        }

        paths
    }

    /// Read file contents with safety checks.
    pub fn read_file(&self, rel_path: &Path) -> Result<String, crate::error::CodeIndexError> {
        use crate::error::CodeIndexError;

        let full = self.config.workspace_root.join(rel_path);

        if !full.exists() {
            return Err(CodeIndexError::FileNotFound(rel_path.to_path_buf()));
        }

        let meta = std::fs::metadata(&full)?;
        if meta.len() > self.config.max_file_size {
            return Err(CodeIndexError::FileTooLarge(
                meta.len(),
                self.config.max_file_size,
            ));
        }

        if Self::is_binary(&full) {
            return Err(CodeIndexError::BinaryFile(rel_path.to_path_buf()));
        }

        if !self.config.follow_symlinks
            && Self::is_symlink_escape(&full, &self.config.workspace_root)
        {
            return Err(CodeIndexError::SymlinkEscape(rel_path.to_path_buf()));
        }

        std::fs::read_to_string(&full).map_err(CodeIndexError::Io)
    }
}
